### PR TITLE
Add basic Xamarin.Mac support

### PR DESCRIPTION
**Pull request by chamons**
- Does not contain "magic" found in YogaKit yet, but enough to get started
- Simple test project showing use
- YGInteropSetLogger and YGNodeInit(IntPtr) were missing from native lib built by buck

### CODE REVIEW

**Comment by chamons** [comment]
CLA should be signed now.

**Comment by emilsjolander** [comment]
@splhack Can you have a look?

File: csharp/Facebook.Yoga/Native.cs [hunk 1]
```diff
@@ -14,21 +14,25 @@ namespace Facebook.Yoga
 {
     internal static class Native
     {
-#if UNITY_IOS && !UNITY_EDITOR
+#if (UNITY_IOS && !UNITY_EDITOR) || MONOMAC
```
Comment (splhack) [inline csharp/Facebook.Yoga/Native.cs]: You can now build libmono.dylib via `buck build //csharp:yoganet#default,shared`. Can you revert `MONOMAC` changes in csharp/Facebook.Yoga/Native.cs, csharp/Facebook.Yoga/YogaLogger.cs, csharp/Facebook.Yoga/YogaNode.cs?

File: csharp/Facebook.Yoga/mac/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+libyoga.a
```
Comment (splhack) [inline csharp/Facebook.Yoga/mac/.gitignore]: libyoga.dylib

File: csharp/Facebook.Yoga/mac/ApiDefinition.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+
+using AppKit;
+using Foundation;
+using ObjCRuntime;
+using CoreGraphics;
+
+namespace Facebook.Yoga.Mac
+{
+	// We're actually abusing the Xamarin.Mac binding project
```
Comment (splhack) [inline csharp/Facebook.Yoga/mac/ApiDefinition.cs]: Apparently we need empty ApiDefinition.cs but don't bind static lib. Can you update the comments?

File: csharp/Facebook.Yoga/mac/CustomBuildAction.targets [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+	<PropertyGroup>
+	    <CompileDependsOn>CopyInNativeLib;$(CompileDependsOn)</CompileDependsOn>
+	</PropertyGroup>
+
+	<!-- The # and , in this path does not play nice with the binding project logic, so make a copy -->
+	<Target Name="CopyInNativeLib" Inputs="..\..\..\buck-out\gen\yoga#default,static\libyoga.a" Outputs="$(ProjectDir)libyoga.a">
```
Comment (splhack) [inline csharp/Facebook.Yoga/mac/CustomBuildAction.targets]: <Target Name="CopyInNativeLib" Inputs="..\..\..\buck-out\gen\csharp\yoganet#default,shared\libyoga.dylib" Outputs="$(ProjectDir)libyoga.dylib">
            <Copy SourceFiles="..\..\..\buck-out\gen\csharp\yoganet#default,shared\libyoga.dylib" DestinationFiles="$(ProjectDir)/libyoga.dylib" />

File: csharp/Facebook.Yoga/mac/Facebook.Yoga.Mac.csproj [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    </Compile>
+    <Compile Include="..\MeasureOutput.cs">
+      <Link>MeasureOutput.cs</Link>
+    </Compile>
+    <Compile Include="..\MeasureFunction.cs">
+      <Link>MeasureFunction.cs</Link>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <NativeReference Include="libyoga.a">
```
Comment (splhack) [inline csharp/Facebook.Yoga/mac/Facebook.Yoga.Mac.csproj]: libyoga.dylib

File: csharp/Facebook.Yoga/mac/Facebook.Yoga.Mac.csproj [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      <Kind>Static</Kind>
```
Comment (splhack) [inline csharp/Facebook.Yoga/mac/Facebook.Yoga.Mac.csproj]: Dynamic

**Comment by chamons** [comment]
All of the icon images are deleted (default project added them) and I swapped over to the dylib. I also handled all of the code review comments except the request to rewrite the sample.

I'll have to spend more time to do that, later.

**Comment by splhack** [comment]
Also can you take a look at my directory hierarchy proposal? https://github.com/facebook/yoga/pull/283#issuecomment-267533985

**Comment by chamons** [comment]
Sample converted. There is one special case hack in the ApplyYogaLayout helper function, but the entire thing is a bit of a hack.

UIView+Yoga.m really needs to get ported over to AppKit.

**Comment by chamons** [comment]
@splhack Anything else for this?

**Comment by emilsjolander** [comment]
@chamons It is holiday season so this might take a while to get fully reviews and merged 👍

**Comment by chamons** [comment]
@emilsjolander Beyond no problem. Just wanted to make sure my end had a nice bow before I disappeared for 🎄

**Comment by splhack** [comment]
@chamons Sorry for the late reply. Do you mind to move Xamarin.Mac related files to `csharp/Xamarin.Mac` directory like #280 `csharp/Xamarin.iOS` directory structure?

**Comment by chamons** [comment]
@splhack I 👍 that PR earlier.

I'm fine w\ that going in and then reworking on top. I think since it's moving stuff around it should go first anyway.

**Comment by emilsjolander** [comment]
Want to make sure you are aware of changes being made to the iOS wrapper currently as it might effect this PR https://github.com/facebook/yoga/pull/289

**Comment by emilsjolander** [comment]
https://github.com/facebook/yoga/pull/283 has been merged now

**Comment by chamons** [comment]
@splhack @emilsjolander Ok, I've:

- Merged master 
- Moved everything to csharp/Mac
- Now consume the shared project instead of defining the file list myself

On the todo list (outside this PR) is porting YogaKit to Mac/Xamarin.Mac.

Anything else needed for this?

**Comment by splhack** [comment]
@chamons Thank you for working on this! looks good to me, just a little thing in coding style, we already have 4 spaces indent in all C# files. Do you mind to replace tab?

**Comment by chamons** [comment]
@splhack Nope. I'll need to remember to teach XS / vim that next time I hack on things.

Let me reformat that...

**Comment by chamons** [comment]
Used: expand -t 4 foo | sponge foo

I think I got them all. I did not both formatting the machine generated files (csproj, sln, info.plist, designer.cs).

**Pull request closed without merging**